backend/src/services/v2/inference.ts

```diff
@@ -0,0 +1,114 @@
+  const model = await getModelById(user, modelId)
+  const inference = await Inference.findOne({
+    modelId,
+    image,
+    tag,
+  })
+
+  if (!inference) {
+    throw NotFound(`The requested inferencing service was not found.`, { modelId, image, tag })
+  }
+
+  const auth = await authorisation.model(user, model, ModelAction.View)
+  if (!auth.success) {
+    throw Forbidden(auth.info, { userDn: user.dn, modelId })
+  }
+  return inference
+}
+
+export async function getInferencesByModel(user: UserDoc, modelId: string) {
+  const model = await getModelById(user, modelId)
+
+  if (!model) {
+    throw NotFound(`The requested model was not found.`, { modelId })
+  }
+
+  const inferences = await InferenceModel.find({ modelId })
+
+  return inferences
+}
+
+export type CreateInferenceParams = Pick<InferenceInterface, 'image' | 'tag' | 'description' | 'settings'>
+
+export async function createInference(user: UserDoc, modelId: string, inferenceParams: CreateInferenceParams) {
+  const model = await getModelById(user, modelId)
+
+  const auth = await authorisation.model(user, model, ModelAction.Update)
+  if (!auth.success) {
+    throw Forbidden(auth.info, {
+      userDn: user.dn,
+      modelId: modelId,
+      image: inferenceParams.image,
+      tag: inferenceParams.tag,
+    })
+  }
+
+  // Check that an image exists in the registry
+  const images = await listModelImages(user, modelId)
+
+  const image = images.find((image) => image.repository === modelId && image.name === inferenceParams.image)
+
+  if (image?.tags.indexOf(inferenceParams.tag) === -1) {
+    throw NotFound(`Image ${modelId}:${inferenceParams.tag} was not found on this model.`, {
+      modelId: modelId,
+    })
+  }
+  const inference = new Inference({
+    modelId: modelId,
+    createdBy: user.dn,
+    ...inferenceParams,
+  })
+
+  try {
+    await inference.save()
+  } catch (error) {
+    if (isMongoServerError(error) && error.code === 11000) {
+      throw BadReq(`A service with this image already exists.`, {
+        modelId: modelId,
+        image: inferenceParams.image,
+        tag: inferenceParams.tag,
+      })
+    }
+
+    throw error
+  }
+
+  return inference
+}
+
+export type UpdateInferenceParams = Pick<InferenceInterface, 'image' | 'tag' | 'description' | 'settings'>
+export async function updateInference(
+  user: UserDoc,
+  modelId: string,
+  inferenceParams: UpdateInferenceParams,
+): Promise<InferenceDoc> {
+  const model = await getModelById(user, modelId)
+
+  const inference = await getInferenceByImage(user, modelId, inferenceParams.image, inferenceParams.tag)
+  const auth = await authorisation.model(user, model, ModelAction.Update)
+
+  if (!auth.success) {
+    throw Forbidden(auth.info, { userDn: user.dn, modelId: modelId })
+  }
+
+  const updatedInference = await Inference.findOneAndUpdate(
+    { modelId: inference.modelId, image: inference.image, tag: inference.tag },
```

Comment:
What would happen here if you tried to update the inference to be the same as an existing one?

Reply:
Also, should you allow the user to update the model ID on the inference setting? You've only checked their permissions on the existing model and the route is specific for the existing model ID. I would assume that they would have to delete the existing one and make another request to create an inference service on another model.